### PR TITLE
Add Statement property to $MyInvocation

### LINKED ISSUE #12609: MyInvocation.Line doesn't work for commands split across multiple lines
My team uses powershell scripts a lot as part of our Azure DevOps pipelines and there is a lot of times we don't have enough information to debug when something goes wrong so I'm been trying to find a nice easy way to output the exact command line so we can try and reproduce it locally. I found an old PS blog post https://devblogs.microsoft.com/powershell/getting-all-your-parameters/ which I thought would do the job so I started using [MyInvocation.Line](https://github.com/Azure/azure-sdk-tools/commit/80de19bbcd4226c37c0085036bdf15b1022afd0c#diff-bb4f526089847eee6eb5471ec830d63c) but the first time I needed to use this information to debug I realized it only contained the first line of the command line and since we often split long command lines across multiple lines to make reading easier that didn't end up helping like I wanted. 

# Steps to reproduce

```powershell
> function test { $myinvocation.line.substring($myInvocation.OffSetInLine-1) }
> test 1 2 3 `
>> 4 5 6
```

# Expected behavior

```none
test 1 2 3 4 5 6
```
or perhaps
```none
test 1 2 3 `
4 5 6
```

# Actual behavior

```none
test 1 2 3 `
```

# Environment data

```none
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.0.0
PSEdition                      Core
GitCommitId                    7.0.0
OS                             Microsoft Windows 10.0.18363
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

I'd also be happy with another simple way to echo the exact command line of a script call.

### PR DESCRIPTION
**Pull request by IISResetMe**
# PR Summary

Fixes #12609

It's currently near-impossible to obtain the exact invocation statement of the current callsite in a script block or script function if the invocation spans multiple lines:

```
PS ~> function Capture-Extent { return $MyInvocation.Line.Substring($MyInvocation.OffSetInLine - 1) }
PS ~> $ext = Capture-Extent @'
these are not the droids you're looking for
'@
PS ~> $ext
Capture-Extent @'
PS ~>
```

This PR exposes the full text of the current invocation via a new instance property `Statement`, allowing command authors to capture multi-line invocations:

```
PS ~> function Capture-Extent { return $MyInvocation.Statement }
PS ~> $ext = Capture-Extent @'
Now we can see what's going on down here!
'@
PS ~> $ext
Capture-Extent @'
Now we can see what's going on down here!
'@
PS ~>
```

## PR Context

WG-Engine previously [discussed this behavior](https://github.com/PowerShell/PowerShell/issues/12609#issuecomment-1209090602) and found no good reason not to expose the full text, but decided against modifying any existing behavior, hence this additional property.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/InvocationInfo.cs [hunk 1]
```diff
@@ -271,6 +271,18 @@ public string Line
...
 
+        /// <summary>
+        /// The full text of the invocation statement, may span multiple lines.
+        /// </summary>
+        /// <value>Statement that was entered to invoke this command</value>
+        public string Statement
+        {
+            get
+            {
+                return ScriptPosition.Text ?? string.Empty;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/InvocationInfo.cs:282]: I think `Text` is never null.

File: src/System.Management.Automation/engine/InvocationInfo.cs [hunk 2]
```diff
@@ -271,6 +271,18 @@ public string Line
             }
         }
 
+        /// <summary>
+        /// The full text of the invocation statement, may span multiple lines.
+        /// </summary>
+        /// <value>Statement that was entered to invoke this command</value>
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/InvocationInfo.cs:277]:
```suggestion
        /// <value>Statement that was entered to invoke this command.</value>
```

File: src/System.Management.Automation/engine/InvocationInfo.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/InvocationInfo.cs:282]: I looked all callsites where the Text is used - we never check for null.

File: src/System.Management.Automation/engine/InvocationInfo.cs [hunk 3]
```diff
@@ -271,6 +271,18 @@ public string Line
...
+        /// The full text of the invocation statement, may span multiple lines.
+        /// </summary>
+        /// <value>Statement that was entered to invoke this command.</value>
+        public string Statement
+        {
+            get
+            {
+                return ScriptPosition.Text;
+            }
+        }
```
Comment (ArmaanMcleod) [inline src/System.Management.Automation/engine/InvocationInfo.cs:284]:
```suggestion
        public string Statement
        {
            get => ScriptPosition.Text;
        }
```

Comment (ArmaanMcleod) [reply src/System.Management.Automation/engine/InvocationInfo.cs:284]: Simplified getter probably not consistent with rest of the file 😄

Comment (IISResetMe) [reply src/System.Management.Automation/engine/InvocationInfo.cs:284]: I deliberately kept style choices consistent with the existing code, yes ^^

**Review by SeeminglyScience — APPROVED** [review]
LGTM! Thank you @IISResetMe ❤️

**Review by adityapatwardhan — APPROVED** [review]

**Comment by iSazonov** [comment]
Do we need new doc issue?

**Comment by daxian-dbw** [comment]
The xml comment block will generate corresponding doc content for the new property in our doc pipeline, so I think we are good.

**Pull request merged**